### DISCUSSION TITLE
Script optimization through compilation

### DISCUSSION THREAD

**Discussion by rsmmr** [zeek/zeek#1203] [Projects]
Creating a ticket to track discussion of [Vern's branch](https://github.com/zeek/zeek/tree/topic/vern/script-opt) adding a script compiler to speed up execution. The goal is to merge this in as an experimental feature, ideally with 4.0 if timing works out.

Some initial thoughts & discussion on how to approach the merge (with a focus on the integration into the current code base) are in this [Google doc](https://docs.google.com/document/d/1EhgR80BkWgIeHpgxWiXegQfcOT6zfQagpvc8LDjFigU).

**Comment by jsiwek** [comment]
After reviewing only the READMEs, it sounds like it's shaping up great and had the following thoughts for what may warrant fleshing out even before an experimental phase.  Some are duplicate/similar to Robin's Google Doc.  Expect I can find answers to questions once I review/investigate further, but maybe @vpax can easily answer some and/or extend the READMEs to cover.

## Testing / CI

- Thinking a requirement before any merge is to have full test suite passing
  in optimized-mode along with CI configured to now run against both modes.

- Hope there can be tests to verify correctness of various optimizations.
  Thinking `btest` along with dumping out the reduced AST or ZAM
  representations/descriptions associated with an optimized-script may be a
  natural approach.  More coverage before even an experimental release would
  be nice.

## Known Deficiencies

- Race condition on writing same ZAM file from many Zeek processes seems worth
  fixing before any merge (very likely people will encounter).

- Wonder what exactly goes wrong with the `when` expressions and
  concurrent/recursive `for` loop execution.  I can see those blocking even an
  experimental release if the user only will see surprising behavior without
  any warning when accidentally wandering into those situations.

## Known Incompatibilities

- Thinking these are all acceptable for experimental phase.

- Long-term, I wonder how burdensome it is to have inconsistencies between
  the two modes.  Any ideas/plans to reduce differences ?  Anything in
  interpreted-mode to deprecate and align to behave the same compiled-mode ?
  Or the other way: anything in compiled-mode that can be made to conform to
  interpreted-mode behavior with some effort ?

## ZAM Files

- Any concern for (hash) collisions ?

  - If so, maybe ZAM file format contains list of all script file dependencies
    with their modification-times and compare against that list.  Also
    dependence on `zeek` binary would be implied.

  - What about the (unlikely, but still valid) case where there's two
    functions / handlers on the same line ?  Do those currently map to the
    same ZAM file ?  Add a starting-column-number in file name to resolve ?

## "Usage Error" Analysis

- In what cases may `&is_set` be truly needed ?  Seems like a good way
  for human error/mistake to sneak in -- either on first setting the
  attribute or during future code maintenance/transformation.  This
  seems important/mandatory for users to get right and for the optimizer
  to be certain about as otherwise the generated code is incorrect ?

- Should the warnings be on by default and require user to opt-out ?
  An "uninitialized usage" always seems worth knowing as it's a bug.  The
  "unused assignment" also always seems helpful and could even help catch dumb
  mistakes (e.g. copy-paste error) that turn out to be a real bug in
  surrounding logic upon reviewing and not only related to optimization.

## Internals / Misc.

- What's the reason for not currently handling/combining `event` and `hook`
  bodies during inlining step ?

**Reply by vpax** [reply]
> * Thinking a requirement before any merge is to have full test suite passing in optimized-mode along with CI configured to now run against both modes.

Makes sense ... but beyond my `btest` fu to know how to set up.

**Reply by vpax** [reply]
> * Hope there can be tests to verify correctness of various optimizations. Thinking `btest` along with dumping out the reduced AST or ZAM representations/descriptions associated with an optimized-script may be a natural approach.  More coverage before even an experimental release would be nice.

Agree, happy to pursue once I understand how to use `btest` variants.

**Reply by vpax** [reply]
> * Race condition on writing same ZAM file from many Zeek processes seems worth fixing before any merge (very likely people will encounter).

Yep.  Robin suggested two possibilities, either per-PID ZAM file generation, or generate-to-temp-file-and-move (since moves are atomic).  I like the latter as it's simple and less likely to lead to a lot of redundancy.  Seems quite simple to implement.

**Reply by vpax** [reply]
> * Wonder what exactly goes wrong with the `when` expressions and concurrent/recursive `for` loop execution.  I can see those blocking even an experimental release if the user only will see surprising behavior without any warning when accidentally wandering into those situations.

The first of these doesn't generate any surprises, the function with the `when` just gets interpreted rather than compiled (invisible to the user, other than potentially becoming more of a performance bottleneck).  This is just because I don't fully understand what sort of interpreter stack frame is needed for `when` and what's the ZAM frame equivalent.  Perhaps it's easy to fix, which would be great.

The `for` issue is conceptually not hard to fix, it's just one that I noticed in late in the game and haven't yet gotten around to (since in practice it hasn't created issues ... which I realize is _not_ an excuse!)

**Reply by vpax** [reply]
> * Long-term, I wonder how burdensome it is to have inconsistencies between the two modes.  Any ideas/plans to reduce differences ?  Anything in interpreted-mode to deprecate and align to behave the same compiled-mode ?

Well, my inclination/prejudice is to over time get rid of interpreted-mode ...

>   Or the other way: anything in compiled-mode that can be made to conform to
>   interpreted-mode behavior with some effort ?

The differences that I'm aware of are due to performance considerations, so hard to give up on given the intent behind the functionality.  Happy though to talk about specific differences!

**Reply by vpax** [reply]
> Any concern for (hash) collisions ?

I wouldn't expect so.  My view is that if there's a collision, that's a bug that should be fixed (by folding more into the hash).  The hash itself is reasonably large (128 bits) and taken from Boost, which I'd hope managed to devise a pretty good (+ cheap) hash function.

**Reply by vpax** [reply]
> * If so, maybe ZAM file format contains list of all script file dependencies with their modification-times and compare against that list.

I'm not sure this is easy to figure out (all dependencies), but happy to fold in if needed.

> Also dependence on `zeek` binary would be implied.

Makes sense, though I also like the notion that it's routine to just blow away all the ZAM files and regenerate them whenever there's any doubt ...

**Reply by vpax** [reply]
> * What about the (unlikely, but still valid) case where there's two functions / handlers on the same line ?  Do those currently map to the same ZAM file ?  Add a starting-column-number in file name to resolve ?

Yeah, I thought about that.  It's only an issue if they also have the same hash ... which seems quite perverse.  I liked not clogging up the filename with column ... though also considered only adding starting-column if it's not 1.

**Reply by vpax** [reply]
> * In what cases may `&is_set` be truly needed ?  Seems like a good way for human error/mistake to sneak in -- either on first setting the attribute or during future code maintenance/transformation.  This seems important/mandatory for users to get right and for the optimizer to be certain about as otherwise the generated code is incorrect ?

You'd think so, but what happens is that there are places where surrounding logic guarantees that a value will be set even though the static analysis can't figure that out.  Here's a _potential_ such place: https://github.com/zeek/zeek/blob/a2c139f405ec9246d9398b11f59670fc350da1aa/scripts/base/protocols/ftp/utils-commands.zeek#L122 (for which I haven't verified that in fact the value will be set, but using `-uu` flags it as potentially not set).

So the point of the attribute is to potentially allow decorating of the base scripts as needed so that any `-uu` complaint that surfaces will get flagged.

**Reply by vpax** [reply]
> * Should the warnings be on by default and require user to opt-out ? An "uninitialized usage" always seems worth knowing as it's a bug.

The problem is that sometimes there's surrounding logic that makes it safe ... _and_ it's expensive to compute, especially for record fields.

>   "unused assignment" also always seems helpful and could even help catch dumb mistakes (e.g. copy-paste error) that turn out to be a real bug in surrounding logic upon reviewing and not only related to optimization.

See the discussion at https://github.com/zeek/zeek/blob/a2c139f405ec9246d9398b11f59670fc350da1aa/scripts/base/frameworks/intel/main.zeek#L476 for why this might be useful.

**Reply by vpax** [reply]
> * What's the reason for not currently handling/combining `event` and `hook` bodies during inlining step ?

Nothing deep, I just was concerned I would screw them up because of the need to correctly chain from one function body to the next.  I agree this would be nice to address.

**Reply by vpax** [reply]
> Yep. Robin suggested two possibilities, either per-PID ZAM file generation, or generate-to-temp-file-and-move (since moves are atomic). I like the latter as it's simple and less likely to lead to a lot of redundancy. Seems quite simple to implement.

I'm gearing up to do this, but wanted feedback on a couple of technical specifics.  First, IIRC there's no `move()` system call, but there's `link()`, which can be followed by `unlink()` - is that right?  Second, I believe this only works if the links are on the same physical partition, yes?  Third, if so then seems I can't reliably use `/tmp` but instead need to use the current directory (where the ZAM file will be written) ... in which case I need to manually make the temporary name unique, say by adding in the hostname and the PID.  Other suggestions?

**Reply by timwoj** [reply]
You might try the `rename()` system call too, but I'm not sure what the restrictions are on it off the top of my head.

**Reply by jsiwek** [reply]
@vpax `mkstemp()` and `rename()` within the same dir as destination/target file is what I think of first.  Otherwise, a `rename()` across filesystems wouldn't work as it would have to instead copy (which is then no longer atomic) plus also don't think a hardcoded `/tmp` path would be reliable either.

**Reply by vpax** [reply]
I've now implemented this approach, seems to be working okay - thanks!

**Reply by rsmmr** [reply]
(deleted comment I made before seeing more replies :)

**Reply by vpax** [reply]
> The `for` issue is conceptually not hard to fix, it's just one that I noticed in late in the game and haven't yet gotten around to (since in practice it hasn't created issues ... which I realize is _not_ an excuse!)

An update: I've now pushed a fix addressing this (including a btest, which I put in a new `opt/` subtree).

**Reply by vpax** [reply]
> > * What about the (unlikely, but still valid) case where there's two functions / handlers on the same line ?  Do those currently map to the same ZAM file ?  Add a starting-column-number in file name to resolve ?
> 
> Yeah, I thought about that. It's only an issue if they also have the same hash ... which seems quite perverse. I liked not clogging up the filename with column ... though also considered only adding starting-column if it's not 1.

Oh, and if they have the same hash, then assuming the hash function is sound (see prior item), then it's _okay_ that they collide, since they should then have identical `.ZAM` code anyway.

**Comment by vpax** [comment]
How do we migrate/link this to Discussions?

**Reply by rsmmr** [reply]
@vpax I just converted it over (there's a button for that in tickets now).

**Reply by vpax** [reply]
Thanks ... not sure I used it correctly above though, as it still seems pretty monolithic ....

**Comment by vpax** [comment]
Oh, and: a great set of points, thanks!

**Comment by rsmmr** [comment]
Extracting a comment by @vpax  from the PR  for discussion: [over here](https://github.com/zeek/zeek/pull/1231#discussion\_r510463705]) for discussion:

> One approach would be to change all of these low-level representations to be `Obj`'s, though getting this right w/o introducing leaks would likely be a pain. If we made that change, then `ZAMValUnion` could be expressed as some additional fields to the existing (modified) `BroValUnion` rather than a separate representation.

A `std::shared_ptr<T>` would give us reference counting without having to derive from `Obj`, so that could be part of a solution.
 
>  Another approach is to get rid of `BroValUnion` and hoist its different variants into the different `Val` subclasses that use them. For example, `RecordVal`'s would have the `PDict` as an explicit member variable, rather than using the common `BroValUnion` to store them. The scope of this change also gives me pause, though, and it could wind up being clunky or potentially buggy for closely-related `Val` subclasses like `TimeVal` and `DoubleVal`. If we did this then we'd still have `ZAMValUnion` but it would be the only such low-level representation of values (other than those spread across different `Val` subclasses).

Mulling over this,  this strikes me as a potentially viable route:
1. Architecturally, this feels generally like a beneficial restructure to me: locating state with the owner instead of a publicly accessible member of the base class, provides encapsulation and type safety.
2. I believe this would help us to contain ZAM-specific logic to just a few well-defined places.
3. I also believe that the fall-out from such a restructure wouldn’t be too huge, and probably pretty mechanical to address.

I admit I still have trouble fully following semantics/properties/goals of some of the pieces in `ZVal.h`, but let me throw out a high-level straw man of this and see if that makes any sense to you:

1. We would get rid of `BroValUnion` and relocate that state into the `Val`-derived classes. This might mean introducing new classes for types that don’t have a `Val`-derived class yet, not sure; if so, that could probably be templated to avoid code duplication.
2. Each of those classes would not provide any raw access to its state; all manipulation must be through methods.
 3. We add one new virtual method to `Val`  that converts into a `ZAMValUnion`; and also a new constructor to each derived class `T` to create it from a `ZAMValUnion`. (Btw: I’d skip the `Union` in that name, and just call it `ZAMValue`, to avoid designing in implementation details):

	```
	# Constructs a ZAM value from the current Val instance.
	class Val { virtual ZAMValUnion ToZAMVal() const = 0; }
	
	# Construct an instance of T from a ZAM value; will be called only for ZAM values of a matching type.
	class T { T(ZAMValUnion&& zv) {...} };
	```

	In addition, we’d have a static factory method like this:

	```
	# Turn an arbitrary ZAM value into a ValPtr.
	static ValPtr Val::FromZAMVal(ZAMValUnion zv) {
		if ( zv already has a valptr )
			return valptr;
		else
			return make_intrusive<type_matching_zv>(std::move(zv));
	}
	```

4. We avoid use of `ZAMValUnion` almost completely outside anything not part of the optimizer, with the only exceptions being these conversion methods.

5.  Generally, the  conversion methods would just do the obvious explicit conversion/instantiation.  Couple additional pieces:

	* This is where the `shared_ptr<T>_` could come in: if we store the state that way, we could just copy that over into a `ZAMValUnion` to share the state.
	* If for performance , we want the option to support different ways of storing aggregate values without explicitly converting between them (i.e., as either `ValPtrs` or `ZAMValUnion`), we could now have a little union/variant internally here. For example,  a `VectorVal` could store either a `vector<ValPtr>` or a `vector<ZAMValUnion>`. As long as we can make usage of all the class’ methods transparent to that internal distinction, caller’s wouldn’t need to care. We can make that all internally safe as well, and it would all be nicely unit-testable too.

6. Not sure what this restructure would mean for `ZAMValUnion`, maybe it could now just store a `ValPtr` instead of enumerating all the individual types (but this is also where I’m getting lost between when we store pointers to Val-derived classes vs. managed values, and who is creating what).

**Reply by rsmmr** [reply]
One more:

7. I believe this would then also mean that we don't need `ZAM_{vector, record}' anymore.

**Reply by vpax** [reply]
I'm game to take a run at getting rid of `BroValUnion` by hoisting its elements into the corresponding `Val` objects.  This will turn some low-level simple inlined functions into virtual functions, so with a potential (minor) performance impact.  I figure you're okay with that, but thought I'd flag to make sure we're all on the same page.

**Reply by rsmmr** [reply]
Sounds good. I don't immediately see what methods turn virtual, but seems fine where they are needed. I would expect some new `dynamic_cast<X>` as well for down-casting from `Val` to, say, a RecordVal, but not really different from the current `AsX()``.
